File: fs/before_energy/297__serialize-and-deserialize-binary-tree/cpp/serialize-and-deserialize-binary-tree_benchmark_v2.py
# Time:  O(n)
# Space: O(h)

class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Codec(object):

    def serialize(self, root):
        """Encodes a tree to a single string.

        :type root: TreeNode
        :rtype: str
        """
        def serializeHelper(node):
            if not node:
                vals.append('#')
                return
            vals.append(str(node.val))
            serializeHelper(node.left)
            serializeHelper(node.right)
        vals = []
        serializeHelper(root)
        return ' '.join(vals)


    def deserialize(self, data):
        """Decodes your encoded data to tree.

        :type data: str
        :rtype: TreeNode
        """
        def deserializeHelper():
            val = next(vals)
            if val == '#':
                return None
            node = TreeNode(int(val))
            node.left = deserializeHelper()
            node.right = deserializeHelper()
            return node
        def isplit(source, sep):
            sepsize = len(sep)
            start = 0
            while True:
                idx = source.find(sep, start)
                if idx == -1:
                    yield source[start:]
                    return
                yield source[start:idx]
                start = idx + sepsize
        vals = iter(isplit(data, ' '))
        return deserializeHelper()


# time: O(n)
# space: O(n)

class Codec2(object):

    def serialize(self, root):
        """Encodes a tree to a single string.
        
        :type root: TreeNode
        :rtype: str
        """
        def gen_preorder(node):
            if not node:
                yield '#'
            else:
                yield str(node.val)
                for n in gen_preorder(node.left):
                    yield n
                for n in gen_preorder(node.right):
                    yield n
                
        return ' '.join(gen_preorder(root))
        
    def deserialize(self, data):
        """Decodes your encoded data to tree.
        
        :type data: str
        :rtype: TreeNode
        """
        def builder(chunk_iter):
            val = next(chunk_iter)
            if val == '#':
                return None
            node = TreeNode(int(val))
            node.left = builder(chunk_iter)
            node.right = builder(chunk_iter)
            return node
        
        # https://stackoverflow.com/a/42373311/568901
        chunk_iter = iter(data.split())
        return builder(chunk_iter)


if __name__ == "__main__":
    tests = [
        "1 # # ",
        "2 1 # # 3 # # ",
        "4 3 2 1 # # # # # ",
        "1 # 2 # 3 # 4 # # ",
        "1 2 4 # # 5 # # 3 6 # # 7 # # ",
        "-10 -20 -30 # # -5 # # 15 0 # # 22 # # ",
        "1 1 1 # # # 1 1 # # 1 # # ",
        "8 4 2 1 # # 3 # # 6 5 # # 7 # # 12 10 9 # # 11 # # 14 13 # # 15 # # ",
        "5 2 # 3 # # 9 # 10 8 # # # ",
        "1 # 2 3 # 4 5 # # # # "
    ]

    # Run tests with Codec (strip trailing spaces to match its parser)
    codec = Codec()
    iterations = 1000
    sink = 0
    for iter in xrange(iterations):
        #checksum = 0
        for s in tests:
            root = codec.deserialize(s.rstrip())
            s2 = codec.serialize(root)
            #checksum += len(s2)
        #sink = checksum

    # # Run tests with Codec2
    # codec2 = Codec2()
    # sink2 = 0
    # for iter in xrange(iterations):
    #     checksum = 0
    #     for s in tests:
    #         root = codec2.deserialize(s)
    #         s2 = codec2.serialize(root)
    #     #     checksum += len(s2)
    #     # sink2 = checksum

File: fs/cpp_before_energy/297__serialize-and-deserialize-binary-tree/serialize-and-deserialize-binary-tree_benchmark_v2.cpp
#include <iostream>
template <class T>
static inline void DoNotOptimize(const T& value) {
    asm volatile("" : : "r,m"(value) : "memory");
}

#include <string>
#include <sstream>
#include <vector>

using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

// Time:  O(n)
// Space: O(h)
class Codec {
public:
    // Encodes a tree to a single string.
    string serialize(TreeNode* root) {
        ostringstream out;
        serializeHelper(root, out);
        return out.str();
    }

    // Decodes your encoded data to tree.
    TreeNode* deserialize(string data) {
        istringstream in(data);
        return deserializeHelper(in);
    }

private:
    void serializeHelper(TreeNode* node, ostringstream& out) {
        if (!node) {
            out << "# ";
            return;
        }
        out << node->val << ' ';
        serializeHelper(node->left, out);
        serializeHelper(node->right, out);
    }

    TreeNode* deserializeHelper(istringstream& in) {
        string val;
        if (!(in >> val)) return nullptr;
        if (val == "#") return nullptr;
        TreeNode* node = new TreeNode(stoi(val));
        node->left = deserializeHelper(in);
        node->right = deserializeHelper(in);
        return node;
    }
};

// Time: O(n)
// Space: O(n)
class Codec2 {
public:
    // Encodes a tree to a single string.
    string serialize(TreeNode* root) {
        ostringstream out;
        genPreorder(root, out);
        return out.str();
    }

    // Decodes your encoded data to tree.
    TreeNode* deserialize(string data) {
        istringstream in(data);
        return build(in);
    }

private:
    void genPreorder(TreeNode* node, ostringstream& out) {
        if (!node) {
            out << "# ";
            return;
        }
        out << node->val << ' ';
        genPreorder(node->left, out);
        genPreorder(node->right, out);
    }

    TreeNode* build(istringstream& in) {
        string val;
        if (!(in >> val)) return nullptr;
        if (val == "#") return nullptr;
        TreeNode* node = new TreeNode(stoi(val));
        node->left = build(in);
        node->right = build(in);
        return node;
    }
};

static void freeTree(TreeNode* root) {
    if (!root) return;
    freeTree(root->left);
    freeTree(root->right);
    delete root;
}

volatile int sink = 0;

int main() {
    vector<string> tests = {
        "1 # # ",
        "2 1 # # 3 # # ",
        "4 3 2 1 # # # # # ",
        "1 # 2 # 3 # 4 # # ",
        "1 2 4 # # 5 # # 3 6 # # 7 # # ",
        "-10 -20 -30 # # -5 # # 15 0 # # 22 # # ",
        "1 1 1 # # # 1 1 # # 1 # # ",
        "8 4 2 1 # # 3 # # 6 5 # # 7 # # 12 10 9 # # 11 # # 14 13 # # 15 # # ",
        "5 2 # 3 # # 9 # 10 8 # # # ",
        "1 # 2 3 # 4 5 # # # # "
    };

    Codec codec;

    const int iterations = 1000;
    for (int iter = 0; iter < iterations; ++iter) {
        //int checksum = 0;
        for (const auto& s : tests) {
            TreeNode* root = codec.deserialize(s);
            string s2 = codec.serialize(root);
            //checksum += static_cast<int>(s2.size());
            DoNotOptimize(s2);
            freeTree(root);
        }
        //sink = checksum;
    }

    return 0;
}